src/main/scala/is/hail/annotations/UnsafeAnnotations.scala

```diff
@@ -0,0 +1,402 @@
+package is.hail.annotations
+
+import is.hail.expr._
+import is.hail.utils._
+import is.hail.variant.{AltAllele, Variant}
+import org.apache.spark.sql.Row
+import org.apache.spark.unsafe.Platform
+
+import scala.collection.mutable.ArrayBuffer
+
+object UnsafeAnnotations {
+  def missingBytes(elems: Int): Int = (elems + 31) / 32 * 4
+
+  def arrayElementSize(t: Type): Int = {
+    var eltSize = t.byteSize
+    val mod = eltSize % t.alignment
+    if (mod != 0)
+      eltSize += (t.alignment - mod)
+    eltSize
+  }
+
+  def roundUpAlignment(offset: Int): Int = {
+    val mod = offset & 0x7
+    if (mod != 0)
+      offset + (8 - mod)
+    else
+      offset
+  }
+}
+
+class UnsafeRowBuilder(t: TStruct, sizeHint: Int = 0, debug: Boolean = false) {
```

Comment:
I think adding a little additional abstraction will make all this cleaner.  I suggest:

```
// I'm not sure about alignment guarantees but Long seems safer
final class MemoryBlock(mem: Array[Long]) {
  def loadInt(off: Int): Int = ...
  // same for Byte, Short, Long, Float, Double, BitArray
  def reallocate(): MemoryBlock
}

final class Pointer(mem: MemoryBlock, memOffset: Int) {
  def loadInt(): Int = loadInt(0)
  def loadInt(off: Int): Int = mem.loadInt(off + memOffset)
  // and others.  No reallocate.

  def offset(off: Int): Pointer = new Pointer(mem, memOffset + off)
}

final class MemoryBuffer(mem: MemoryBlock, offset: Int) {
  // expose MemoryBlock operations, plus:
  def appendInt(i: Int) // reallocate if necessary
  // other types...
  def allocate(nBytes: Int): Int // reallocate if necessary, return offset
  def align(alignment: Int)
}
```